src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java

```diff
@@ -0,0 +1,89 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.proxy.callback;
+
+import io.r2dbc.proxy.core.ConnectionInfo;
+import io.r2dbc.proxy.core.MethodExecutionInfo;
+import io.r2dbc.spi.Batch;
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.Statement;
+
+import java.lang.reflect.Method;
+import java.util.function.Consumer;
+
+/**
+ * Proxy callback for {@link Connection}.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class ReactiveConnectionCallback extends CallbackSupport {
+
+    private Connection connection;
+
+    private ConnectionInfo connectionInfo;
+
+    public ReactiveConnectionCallback(Connection connection, ConnectionInfo connectionInfo, ProxyConfig proxyConfig) {
+        super(proxyConfig);
+        this.connection = connection;
+        this.connectionInfo = connectionInfo;
+    }
+
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+        String methodName = method.getName();
+
+        if ("unwrap".equals(methodName)) {
+            return this.connection;
+        } else if ("getOriginalConnection".equals(methodName)) {
+            return this.connection;
+        }
+
+        Consumer<MethodExecutionInfo> onComplete = null;
+
+        // since these methods return Publisher<Void> pass the callback for doOnComplete().
+        if ("beginTransaction".equals(methodName)) {
+            onComplete = executionInfo -> {
+                executionInfo.getConnectionInfo().incrementTransactionCount();
+            };
+        } else if ("commitTransaction".equals(methodName)) {
+            onComplete = executionInfo -> {
+                executionInfo.getConnectionInfo().incrementCommitCount();
+            };
+        } else if ("rollbackTransaction".equals(methodName)) {
+            onComplete = executionInfo -> {
+                executionInfo.getConnectionInfo().incrementRollbackCount();
+            };
+        } else if ("close".equals(methodName)) {
+            onComplete = executionInfo -> {
+                executionInfo.getConnectionInfo().setClosed(true);
+            };
+        }
+        // TODO: createSavepoint, releaseSavepoint, rollbackTransactionToSavepoint
+
+        Object result = proceedExecution(method, this.connection, args, this.proxyConfig.getListeners(), this.connectionInfo, null, onComplete);
+
+        if ("createBatch".equals(methodName)) {
+            return this.proxyConfig.getProxyFactory().createBatch((Batch) result, this.connectionInfo);
+        } else if ("createStatement".equals(methodName)) {
+            String query = (String) args[0];
+            return this.proxyConfig.getProxyFactory().createStatement((Statement) result, query, this.connectionInfo);
+        }
+
```

Comment:
Is keying off every API's method name via String the only way to manage this?

Reply:
Another way would be to compare `Method` object.

However, I think there are drawbacks for `Method` comparison.
- Need to programmatically prepare `Method` objects.
- Handling of overloading methods (same name but different args)

So, even though comparing with String method name ties to literal method name typing, it is much easier to use for controlling logic flow.

Reply:
Using strings is quite a common pattern and the `Method` approach is significantly more difficult, as someone might redeclare a method and then it's a different one than retrieved from the interface.